### DISCUSSION TITLE
Help with calculation problem i guess

### DISCUSSION THREAD

**Discussion by tilenHD** [community/community#22463] [Projects and Issues]
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/8/a/8a60d32e9ac6b94c9bdf8859abceed66ada3bd6b.png" data-download-href="/uploads/short-url/jK9kUSvlWnQr94Tzr5Jyv3MolXJ.png?dl=1" title="Posnetek zaslona 2022-03-09 182104" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/101281576/181071250-80ef0438-cf14-4e4f-aad2-ce1c45774910.png" alt="Posnetek zaslona 2022-03-09 182104" data-base62-sha1="jK9kUSvlWnQr94Tzr5Jyv3MolXJ" width="513" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/8/a/8a60d32e9ac6b94c9bdf8859abceed66ada3bd6b_2_513x500.png, https://aws1.discourse-cdn.com/github/optimized/3X/8/a/8a60d32e9ac6b94c9bdf8859abceed66ada3bd6b_2_769x750.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/8/a/8a60d32e9ac6b94c9bdf8859abceed66ada3bd6b.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/8/a/8a60d32e9ac6b94c9bdf8859abceed66ada3bd6b_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">Posnetek zaslona 2022-03-09 182104</span><span class="informations">966×940 29.1 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div><br>
Im very new to programing just started in school so i think this is a very dumb problem but to be short i dont know why im getting only 0% and 100% when caculating precentege of the x value? that is randomly genereted bettwen 0 and 100 sorry for my bad english im not native.

**Comment by airtower-luna** [comment]
<p>It’s because <code>x</code> is an <code>int</code>. That’s short for <code>integer</code>, so a whole number (no fraction!). The <code>int</code> type in Java has also certain size limits, but that doesn’t matter here.</p>
<p><code>Math.random()</code> returns a <code>double</code>, which can hold fractions. When a <code>double</code> is casted (turned into) an <code>int</code> anything behind the dot is cut off, for example:</p>
<pre><code class="lang-java">int i = (int) 2.6;
</code></pre>
<p>In this case <code>i</code> will actually be 2. With that knowledge look at line 27 and the <a href="https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Math.html#random()"><code>Math.random()</code> documentation</a>, and calculate which values <code>x</code> can have there. :smiley_cat:</p>
<p>As a side note, line 32 doesn’t do anything. Anything divided by 100 times 100 is the original number again.</p>

**Comment by tilenHD** [comment]
<p>for line 32 i just did this formula (value/total value)×100 so value is x divided by the total value that is 100 and multiplied by 100 so if i give x a valuve of 50 and do 50/100*100 i get 50% (i tested on the calculator)</p>
<p>so if i just convert x to double it should it work?</p>
<p>(i like programing and enjoy it a lot but im not that good at math)<br>
but by programing i learn it more :slight_smile:</p>

**Comment by tilenHD** [comment]
<p>it worked now i know that double gives you decimal vaules but i still dont exatcly know why int only gives you 0 or 100 so i have some reading to do thanks for your help man or woman :slight_smile:</p>
<p>-Tilen from Slovenia<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/b/a/baac4efe366ad60e065fa5a3c29802f4c4d22360.png" data-download-href="/uploads/short-url/qDnYGUnomIYB542tci1aadM0QYE.png?dl=1" title="Posnetek zaslona 2022-03-09 190723" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/101281576/181071256-e026e012-5559-4ac0-bfc2-8b471bd48129.png" alt="Posnetek zaslona 2022-03-09 190723" data-base62-sha1="qDnYGUnomIYB542tci1aadM0QYE" width="687" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/b/a/baac4efe366ad60e065fa5a3c29802f4c4d22360_2_687x500.png, https://aws1.discourse-cdn.com/github/optimized/3X/b/a/baac4efe366ad60e065fa5a3c29802f4c4d22360_2_1030x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/b/a/baac4efe366ad60e065fa5a3c29802f4c4d22360_2_1374x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/b/a/baac4efe366ad60e065fa5a3c29802f4c4d22360_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">Posnetek zaslona 2022-03-09 190723</span><span class="informations">1417×1030 51.6 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by tilenHD** [comment]
<p>I just signed onto this website and respond time is crazy well its programing of course people will be on their computers i guess :slight_smile: .</p>

**Comment by airtower-luna** [comment, marked as answer]
<aside class="quote no-group" data-username="tilenHD" data-post="5" data-topic="235197">
<div class="title">
<div class="quote-controls"></div>
 tilenHD:</div>
<blockquote>
<p>but i still dont exatcly know why int only gives you 0 or 100</p>
</blockquote>
</aside>
<p>Think about the math. You had these definitions (ignoring the loop):</p>
<pre><code class="lang-java">int min = 0;
int max = 100;
int x = (int)(Math.random() * ((max - min) + 1)) + min;
int r = x / 100 * 100;
</code></pre>
<p>Insert the numbers and you get this:</p>
<pre><code class="lang-java">int x = (int)(Math.random() * 101)
int r = x / 100 * 100;
</code></pre>
<p><code>Math.random()</code> returns a number from 0 to infinitely less than 1. So multiplying with something you get something from 0 to less than 101.</p>
<p>Where it breaks is the <code>x / 100</code>. It’s an <em>integer</em> division, so any fraction gets cut off. Which means if x is anything <em>less than 100</em>, the result is 0. If it’s 100.something the result is 1. Only the result of that is then multiplied by 100, so r is either 0 or 100.</p>
<p>In regular math x * 100 / 100 = x, but if anything gets cut off it breaks.</p>

**Comment by tilenHD** [comment]
<p>o yeah i get it now math.random uses those decimals to caculate correctly i think my techer even warned us about this.</p>
<p>thanks for explaining.</p>
